### PR TITLE
Generator architecture

### PR DESCRIPTION
**Pull request by vict0rsch**
* **Encoder**: Resnet-based Content Encoder from MUNIT
  * image => 64 (=`encoder.dim`) channels with 1 conv layer, same size
  * conv-based downsamplings (`encoder.n_downsample` times)
  * resblocks (`encoder.n_res` blocks)
* **Decoders**: Resnet-based Decoders from MUNIT for all tasks but the translation
  * resblocks projections (`decoder.n_res` blocks)
  * Sequence of `nn.Upsampling > Conv2dBlock` (`decoder.n_upsample` times)
    * should match `encoder.n_downsample`
  * final conv to get a feature map with 1 (`h`, `d`, `w`), 3 (`a`) or 19 (`s`) channels
* **Translation decoder**: SPADEResnet-based Decoder inspired by MUNIT and SPADE
  * Conditioning the translation by `SPADE([h, d, s, w])`

### CODE REVIEW

File: README.md [hunk 1]
```diff
@@ -1,5 +1,54 @@
...
+Params size (MB): 149.21
+Estimated Total Size (MB): 4401.17
+----------------------------------------------------------------
+```
+
+Set `test_summary` to `True` in `tests/test_gen.py` to view the full summary.
+
+**n.b.**: the adaptation decoder is not taken into account in the summary as its computations are not used in `OmniGenerator.forward(...)` and only one translation decoder is used so numbers above are a lower bound.
+
+### Generator
```
Comment (vict0rsch) [inline README.md]: here's the high level description of what I've done here

File: omnigan/discriminator.py [hunk 2]
```diff
@@ -64,64 +65,6 @@ def init_net(net, init_type="normal", init_gain=0.02, gpu_ids=[]):
     return net
 
 
-def l2normalize(v, eps=1e-12):
```
Comment (vict0rsch) [inline omnigan/discriminator.py]: moved this to blocks.py

File: omnigan/generator.py [hunk 3]
```diff
@@ -1,14 +1,15 @@
...
-# * https://docs.fast.ai/vision.models.unet.html
-# * Dynamic UNet FastAI
+from omnigan.blocks import Conv2dBlock, ResBlocks, SpadeDecoder, BaseDecoder
 
 # --------------------------------------------------------------------------
 # -----  For now no network structure, just project in a 64 x 32 x 32  -----
 # -----   latent space and decode to (3 or 1) x 256 x 256              -----
 # --------------------------------------------------------------------------
 
+# TODO think about how to use the classifier probs at inference
```
Comment (vict0rsch) [inline omnigan/generator.py]: I'll need you on this @adrienju: we've got to brainstorm about how to use the 1-h vectors specifying the target domain to the generator.

It could be that we don't actually need 2 different translation decoders, and probably not even an adaptation decoder as long as we use the "bit-flip" trick. This would mean that with only 1 decoder we can target the 4 domains if we condition the SPADE blocks on   1-hot encoded (or soft distributed with a softmax) feature maps:

vector (size [4,]) * ones (size [4, 256, 256])

makes sense? To be discussed anyway

File: omnigan/generator.py [hunk 4]
```diff
@@ -82,67 +73,156 @@ def __init__(self, opts):
 
         self.decoders = nn.ModuleDict(self.decoders)
 
+    def forward(self, x, translate_to="f", classifier_probs=None):
```
Comment (vict0rsch) [inline omnigan/generator.py]: generator forward, meaning translating

**Comment by vict0rsch** [comment]
@adrienju there's a lot to review here, I'm not expecting you to go into details but if you could simply go over the updates to the code and tell me if anything seems weird to your eyes / what tests might be missing that'd be great
